reconcile/sentry_config.py

```diff
@@ -0,0 +1,317 @@
+
+  def init_projects_from_desired_state(self, projects):
+    self.projects = projects
+
+  def init_teams(self, teams):
+    self.teams = teams
+
+class SentryReconciler:
+  def __init__(self, client, dry_run):
+    self.client = client
+    self.dry_run = dry_run
+
+  def reconcile(self, current, desired):
+    # Reconcile the teams frist
+    for team in current.teams:
+      if team not in desired.teams:
+        logging.info("deleting sentry team %s" % team)
+        if not self.dry_run:
+          self.client.delete_team(team)
+
+    for team in desired.teams:
+      if team not in current.teams:
+        logging.info("creating sentry team %s" % team)
+        if not self.dry_run:
+          self.client.create_team(team)
+
+    # Reconcile users
+    for user in current.users.keys():
+      if user not in desired.users.keys():
+        logging.info("deleting sentry user %s" % user)
+        if not self.dry_run:
+          self.client.delete_user(user)
+
+    for user in desired.users.keys():
+      teams = desired.users[user]
+      if user not in current.users.keys():
+        logging.info("adding sentry user %s with teams %s" % (user, ",".join(teams)))
+        if not self.dry_run:
+          self.client.create_user(user, "member", teams)
+      else:
+        logging.info("setting sentry user %s team membership to %s" % (user, ",".join(teams)))
+        if not self.dry_run:
+          self.client.set_user_teams(user, teams)
+        logging.info("setting user role for %s to %s" % (user, "member"))
+        if not self.dry_run:
+          self.client.change_user_role(user, "member")
+
+    # Reconcile projects
+    for projects in current.projects.values():
+      for current_project in projects:
+        if project_in_project_list(current_project, desired.projects.values()):
+          continue
+        project_name = current_project["name"]
+        logging.info("deleting project %s" % project_name)
+        if not self.dry_run:
+          self.client.delete_project(project_name)
+
+    for team in desired.projects.keys():
+      for desired_project in desired.projects[team]:
+        project_name = desired_project["name"]
+        if not project_in_project_list(desired_project, current.projects.values()):
+          logging.info("adding project %s" % project_name)
+          if not self.dry_run:
+            self.client.create_project(team, project_name)
+        str = ("setting project config to {}").format(desired_project)
+        logging.info(str)
+        if not self.dry_run:
+          self.client.update_project(project_name, desired_project)
+
+
+def project_in_project_list(project, list):
+  for projects in list:
+    for p in projects:
+      if p["name"] == project["name"]:
+        return True
+  return False
+
+def fetch_current_state(client, ignore_users):
+  state = SentryState()
+
+  # Retrieve all the teams
+  sentry_teams = client.get_teams()
+  teams = []
+  for team in sentry_teams:
+    teams.append(team["slug"])
```

Comment:
this is good opportunity for a list comprehension:

```python
teams = [team['slug'] for team in sentry_teams]
```

Reply:
Done